File: frontend/cypress/test-solutions/serialize-and-deserialize-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

class Solution {
    serialize(root) {
        const vals = [];
        
        function preorder(node) {
            if (!node) {
                vals.push("null");
            } else {
                vals.push(node.val.toString());
                preorder(node.left);
                preorder(node.right);
            }
        }
        
        preorder(root);
        return vals.join(',');
    }
    
    deserialize(data) {
        const vals = data.split(',');
        let index = 0;
        
        function build() {
            if (index >= vals.length || vals[index] === "null") {
                index++;
                return null;
            }
            const node = new TreeNode(parseInt(vals[index]));
            index++;
            node.left = build();
            node.right = build();
            return node;
        }
        
        return build();
    }
}

File: frontend/cypress/test-solutions/serialize-and-deserialize-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    
    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        List<String> vals = new ArrayList<>();
        preorder(root, vals);
        return String.join(",", vals);
    }
    
    private void preorder(TreeNode node, List<String> vals) {
        if (node == null) {
            vals.add("null");
        } else {
            vals.add(String.valueOf(node.val));
            preorder(node.left, vals);
            preorder(node.right, vals);
        }
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        String[] vals = data.split(",");
        return build(new int[]{0}, vals);
    }
    
    private TreeNode build(int[] index, String[] vals) {
        if (index[0] >= vals.length || vals[index[0]].equals("null")) {
            index[0]++;
            return null;
        }
        TreeNode node = new TreeNode(Integer.valueOf(vals[index[0]]));
        index[0]++;
        node.left = build(index, vals);
        node.right = build(index, vals);
        return node;
    }
}